src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright (C) Yichun Zhang (agentzh)
+ * Copyright (C) Jinhua Luo (kingluo)
+ * I hereby assign copyright in this code to the lua-nginx-module project,
+ * to be licensed under the same terms as the rest of the code.
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_worker_thread.h"
+#include "ngx_http_lua_util.h"
+
+
+#if (NGX_THREADS)
+
+
+#include <ngx_thread.h>
+#include <ngx_thread_pool.h>
+
+
+typedef struct ngx_http_lua_task_ctx_s {
+    lua_State                        *vm;
+    ngx_thread_mutex_t                mutex;
+    struct ngx_http_lua_task_ctx_s   *next;
+} ngx_http_lua_task_ctx_t;
+
+
+typedef struct {
+    ngx_http_lua_task_ctx_t *ctx;
+    ngx_http_lua_co_ctx_t   *wait_co_ctx;
+    ngx_http_cleanup_pt      cleanup;
+    void                    *cleanup_data;
+    int                      n_args;
+    int                      rc;
+    int                      is_abort;
+} ngx_http_lua_my_thread_ctx_t;
+
+
+static  ngx_http_lua_task_ctx_t   dummy_ctx;
+static  ngx_http_lua_task_ctx_t  *ctxpool = &dummy_ctx;
+
+
+static ngx_http_lua_task_ctx_t *
+ngx_http_lua_get_task_ctx(lua_State *L)
+{
+    ngx_http_lua_task_ctx_t *ctx = NULL;
+
+    size_t           path_len;
+    const char      *path;
+    size_t           cpath_len;
+    const char      *cpath;
+    lua_State       *vm;
+
+    if (ctxpool->next == NULL) {
+        ctx = calloc(sizeof(ngx_http_lua_task_ctx_t), 1);
+        vm = luaL_newstate();
+        ctx->vm = vm;
+        luaL_openlibs(ctx->vm);
+
+        lua_getglobal(L, "package");
+        lua_getglobal(vm, "package");
+
+        /* copy package.path */
+        lua_getfield(L, -1, "path");
+        path = lua_tolstring(L, -1, &path_len);
+        lua_pushlstring(vm, path, path_len);
+        lua_setfield(vm, -2, "path");
+        lua_pop(L, 1);
+
+        /* copy package.cpath */
+        lua_getfield(L, -1, "cpath");
+        cpath = lua_tolstring(L, -1, &cpath_len);
+        lua_pushlstring(vm, cpath, cpath_len);
+        lua_setfield(vm, -2, "cpath");
+        lua_pop(L, 1);
+
+        /* remove the "package" table */
+        lua_pop(L, 1);
+        lua_pop(vm, 1);
```

Comment:
code style: we should put the code lines that are related to `L` together.

we only need to get `path` and `cpath` from `L`, so, separate the `L` and `vm` will make the code much clear.